File: src/models/Player.java
package models;

import abstracts.GameItem;
import main.Main;
import utils.ImageManager;
import utils.TimeManager;
import utils.UserManager;

import java.awt.*;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

public class Player extends GameItem {
    private final String username;
    private int currentHighScore;

    private int livesLeft = 3;
    private int score = 0;

    // Variables for player movement
    private boolean moveUp = false;
    private boolean moveDown = false;
    private boolean moveLeft = false;
    private boolean moveRight = false;

    public Player(String username) {
        super();

        this.username = username;
        this.currentHighScore = UserManager.getHighScoreForUser(username);

        setAsset(ImageManager.load(PLAYER_ASSET_PATH));
        setPosition(PLAYER_STARTING_POSITION); // 420, 620
    }

    public String getUsername() {
        return username;
    }

    public int getCurrentHighScore() {
        return currentHighScore;
    }

    public void setCurrentHighScore(int currentHighScoreSupplier) {
        this.currentHighScore = currentHighScoreSupplier;
    }

    public void setMoveUp(boolean moveUp) {
        this.moveUp = moveUp;
    }

    public void setMoveDown(boolean moveDown) {
        this.moveDown = moveDown;
    }

    public void setMoveLeft(boolean moveLeft) {
        this.moveLeft = moveLeft;
    }

    public void setMoveRight(boolean moveRight) {
        this.moveRight = moveRight;
    }

    @Override
    public void setIsValid() {
        this.isValid = livesLeft > 0;
    }

    @Override
    public double getTravelSpeed() {
        return PLAYER_TRAVEL_SPEED;
    }

    @Override
    public void updatePosition(float deltaTime) { // The position of player will be updated according to the user input
        double speed = getTravelSpeed();

        if (moveUp && getPosition().y - speed >= (double) SCREEN_HEIGHT / 2) {
            getPosition().y -= speed * deltaTime;
        }

        if (moveDown && getPosition().y + speed + asset.getHeight() <= SCREEN_HEIGHT - this.getAsset().getHeight()) {
            getPosition().y += speed * deltaTime;
        }

        if (moveLeft && getPosition().x - speed >= 0) {
            getPosition().x -= speed * deltaTime;
        }

        if (moveRight && getPosition().x + speed + asset.getWidth() <= SCREEN_WIDTH) {
            getPosition().x += speed * deltaTime;
        }
    }

    @Override
    protected void draw(Graphics g) {
        g.drawImage(getAsset(), (int) getPosition().getX(), (int) getPosition().getY(), null);
    }

    public int getLivesLeft() {
        return livesLeft;
    }

    public void setLivesLeft(int livesLeft) {
        this.livesLeft = livesLeft;
    }

    public int getScore() {
        return score;
    }

    public void setScore(int score) {
        this.score = score;
    }

    public void fireBullets() {
        TimeManager.startTimer(BULLET_FIRE_FREQUENCY, e -> new Bullet(this), () -> !isValid());
    }

    // Investigate this
    private void controlKeyPressedActionHandler(KeyEvent e) {
        try {
            switch (e.getKeyCode()) {
                case KeyEvent.VK_W:
                    setMoveUp(true);
                    break;
                case KeyEvent.VK_S:
                    setMoveDown(true);
                    break;
                case KeyEvent.VK_A:
                    setMoveLeft(true);
                    break;
                case KeyEvent.VK_D:
                    setMoveRight(true);
                    break;
            }
        } catch (IllegalStateException illegalStateException) {
            System.err.printf("%s", getLivesLeft());
            illegalStateException.printStackTrace();
        }
    }

    private void controlKeyReleasedActionHandler(KeyEvent e) {
        switch (e.getKeyCode()) {
            case KeyEvent.VK_W:
                setMoveUp(false);
                break;
            case KeyEvent.VK_S:
                setMoveDown(false);
                break;
            case KeyEvent.VK_A:
                setMoveLeft(false);
                break;
            case KeyEvent.VK_D:
                setMoveRight(false);
                break;
        }

        // Only run in debug mode
        if (Main.debug) {
            System.out.printf(
                    "Player location (x, y): %.2f, %.2f\n",
                    getPosition().getX(),
                    getPosition().getY()
            );
        }
    }

    public KeyAdapter handleUserInput() {
        return new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                super.keyPressed(e);
                controlKeyPressedActionHandler(e);
            }

            @Override
            public void keyReleased(KeyEvent e) {
                super.keyReleased(e);
                controlKeyReleasedActionHandler(e);
            }
        };
    }
}

File: src/items/Player.kt
package items

import abstracts.AbstractGameItem
import constants.GameConstants
import engine.ImageManager.load
import engine.TimeManager.startTimer
import engine.UserManager.getHighScoreForUser
import main.Main
import java.awt.Graphics
import java.awt.event.ActionEvent
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent

class Player(@JvmField val username: String) : AbstractGameItem() {
    @JvmField
    var currentHighScore: Int

    @JvmField
    var livesLeft = 3

    @JvmField
    var score = 0

    // Variables for player movement
    private var moveUp = false
    private var moveDown = false
    private var moveLeft = false
    private var moveRight = false

    init {
        currentHighScore = getHighScoreForUser(username)
        asset = load(GameConstants.PLAYER_ASSET_PATH)
        position = GameConstants.PLAYER_STARTING_POSITION // 420, 620
    }

    fun setMoveUp(moveUp: Boolean) {
        this.moveUp = moveUp
    }

    fun setMoveDown(moveDown: Boolean) {
        this.moveDown = moveDown
    }

    fun setMoveLeft(moveLeft: Boolean) {
        this.moveLeft = moveLeft
    }

    fun setMoveRight(moveRight: Boolean) {
        this.moveRight = moveRight
    }

    public override fun setIsValid() {
        isValid = livesLeft > 0
    }

    override fun setIsValid(isValid: Boolean) {
        // ignored
    }

    override val travelSpeed: Double
        get() = GameConstants.PLAYER_TRAVEL_SPEED

    public override fun updatePosition(deltaTime: Float) { // The position of player will be updated according to the user input
        val speed = travelSpeed
        if (moveUp && position.y - speed >= GameConstants.SCREEN_HEIGHT.toDouble() / 2) {
            position.y -= speed * deltaTime
        }
        if (moveDown && position.y + speed + asset!!.height <= GameConstants.SCREEN_HEIGHT - asset!!.height) {
            position.y += speed * deltaTime
        }
        if (moveLeft && position.x - speed >= 0) {
            position.x -= speed * deltaTime
        }
        if (moveRight && position.x + speed + asset!!.width <= GameConstants.SCREEN_WIDTH) {
            position.x += speed * deltaTime
        }
    }

    override fun draw(g: Graphics?) {
        g!!.drawImage(asset, position.getX().toInt(), position.getY().toInt(), null)
    }

    fun fireBullets() {
        startTimer(GameConstants.BULLET_FIRE_FREQUENCY, { e: ActionEvent? -> Bullet(this) }) { !isValid() }
    }

    // Investigate this
    private fun controlKeyPressedActionHandler(e: KeyEvent) {
        try {
            when (e.keyCode) {
                KeyEvent.VK_W -> setMoveUp(true)
                KeyEvent.VK_S -> setMoveDown(true)
                KeyEvent.VK_A -> setMoveLeft(true)
                KeyEvent.VK_D -> setMoveRight(true)
            }
        } catch (illegalStateException: IllegalStateException) {
            System.err.printf("%s", livesLeft)
            illegalStateException.printStackTrace()
        }
    }

    private fun controlKeyReleasedActionHandler(e: KeyEvent) {
        when (e.keyCode) {
            KeyEvent.VK_W -> setMoveUp(false)
            KeyEvent.VK_S -> setMoveDown(false)
            KeyEvent.VK_A -> setMoveLeft(false)
            KeyEvent.VK_D -> setMoveRight(false)
        }

        // Only run in debug mode
        if (Main.debug) {
            System.out.printf(
                    "Player location (x, y): %.2f, %.2f\n",
                    position.getX(),
                    position.getY()
            )
        }
    }

    fun handleUserInput(): KeyAdapter {
        return object : KeyAdapter() {
            override fun keyPressed(e: KeyEvent) {
                super.keyPressed(e)
                controlKeyPressedActionHandler(e)
            }

            override fun keyReleased(e: KeyEvent) {
                super.keyReleased(e)
                controlKeyReleasedActionHandler(e)
            }
        }
    }
}
